hudi-hadoop-common/src/main/java/org/apache/parquet/avro/HoodieAvroReadSupport.java

```diff
@@ -51,6 +51,13 @@ public ReadContext init(Configuration configuration, Map<String, String> keyValu
       configuration.set(AvroWriteSupport.WRITE_OLD_LIST_STRUCTURE,
           "false", "support reading avro from non-legacy map/list in parquet file");
     }
+    // If old file is written with legacy mode, read with legacy mode, even if now it's non-legacy mode in conf.
+    // Since by changing the property users want to control how files are written, not how they are read.
+    // Later the value of WRITE_OLD_LIST_STRUCTURE will be picked from conf, thus either keeping old mode or writing new mode.
+    if (legacyMode) {
+      configuration.set(AvroWriteSupport.WRITE_OLD_LIST_STRUCTURE,
```

Comment:
look like user has set this conf `false` to controller the action, we'd better not change user behavior. just like`datetimeRebaseModeInWrite`, I don't want compatible history when I set `EXCEPTION`.

Reply:
This is what exactly leads to reading `null` from list fields even though they contain a value.
This leads to exception or silent data loss.
Basically - the change is for reading, while this setting is for writing. The writing part still will be used.
e.g. if I set WRITE_OLD_LIST_STRUCTURE to any value - I expect Hudi/Spark to be able to read 2 levels and 3 level while producing only 3 levels, but in the current state - Hudi reads nulls from 2 level parquets if the setting is set to false

Reply:
I think indeed this change requires thoroughful check, I believe experts in this code can find a better solution. At least the test I've written should test the potential other scenarios (for any fix being developed)